File: android/app/src/main/java/my/ssdid/wallet/feature/device/DeviceEnrollScreen.kt
package my.ssdid.wallet.feature.device

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import my.ssdid.wallet.domain.device.DeviceManager
import my.ssdid.wallet.domain.device.PairingData
import my.ssdid.wallet.domain.model.Identity
import my.ssdid.wallet.domain.vault.Vault
import my.ssdid.wallet.ui.theme.*
import javax.inject.Inject

sealed class EnrollState {
    object Idle : EnrollState()
    data class WaitingForSecondary(val pairingData: PairingData) : EnrollState()
    data class PairingJoined(val pairingId: String, val deviceName: String?) : EnrollState()
    object Approved : EnrollState()
    data class JoinSuccess(val status: String) : EnrollState()
    data class Error(val message: String) : EnrollState()
}

@HiltViewModel
class DeviceEnrollViewModel @Inject constructor(
    private val vault: Vault,
    private val deviceManager: DeviceManager,
    savedStateHandle: SavedStateHandle
) : ViewModel() {
    private val keyId: String = savedStateHandle["keyId"] ?: ""
    val mode: String = savedStateHandle["mode"] ?: "primary"

    private val _identity = MutableStateFlow<Identity?>(null)
    val identity = _identity.asStateFlow()

    private val _state = MutableStateFlow<EnrollState>(EnrollState.Idle)
    val state = _state.asStateFlow()

    init {
        viewModelScope.launch { _identity.value = vault.getIdentity(keyId) }
    }

    fun initiatePairing() {
        val id = _identity.value ?: return
        viewModelScope.launch {
            deviceManager.initiatePairing(id)
                .onSuccess { data ->
                    _state.value = EnrollState.WaitingForSecondary(data)
                    pollPairingStatus(id, data.pairingId)
                }
                .onFailure { _state.value = EnrollState.Error(it.message ?: "Failed to initiate pairing") }
        }
    }

    private fun pollPairingStatus(identity: Identity, pairingId: String) {
        viewModelScope.launch {
            repeat(60) {
                delay(3000)
                deviceManager.checkPairingStatus(identity.did, pairingId)
                    .onSuccess { resp ->
                        if (resp.status == "joined") {
                            _state.value = EnrollState.PairingJoined(pairingId, resp.device_name)
                            return@launch
                        }
                    }
            }
            val current = _state.value
            if (current is EnrollState.WaitingForSecondary) {
                _state.value = EnrollState.Error("Pairing timed out")
            }
        }
    }

    fun approvePairing() {
        val id = _identity.value ?: return
        val current = _state.value
        val pairingId = when (current) {
            is EnrollState.PairingJoined -> current.pairingId
            else -> return
        }
        viewModelScope.launch {
            deviceManager.approvePairing(id, pairingId)
                .onSuccess { _state.value = EnrollState.Approved }
                .onFailure { _state.value = EnrollState.Error(it.message ?: "Failed to approve") }
        }
    }

    fun joinPairing(pairingId: String, challenge: String, deviceName: String) {
        val id = _identity.value ?: return
        viewModelScope.launch {
            deviceManager.joinPairing(
                did = id.did,
                pairingId = pairingId,
                challenge = challenge,
                identity = id,
                deviceName = deviceName
            )
                .onSuccess { status -> _state.value = EnrollState.JoinSuccess(status) }
                .onFailure { _state.value = EnrollState.Error(it.message ?: "Failed to join") }
        }
    }
}

@Composable
fun DeviceEnrollScreen(
    onBack: () -> Unit,
    viewModel: DeviceEnrollViewModel = hiltViewModel()
) {
    val identity by viewModel.identity.collectAsState()
    val state by viewModel.state.collectAsState()
    val isPrimary = viewModel.mode == "primary"

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(BgPrimary)
            .statusBarsPadding()
    ) {
        // Header
        Row(Modifier.padding(20.dp), verticalAlignment = Alignment.CenterVertically) {
            TextButton(onClick = onBack) { Text("\u2190", color = TextPrimary, fontSize = 20.sp) }
            Spacer(Modifier.width(12.dp))
            Text(
                if (isPrimary) "Enroll New Device" else "Join as Device",
                style = MaterialTheme.typography.titleLarge
            )
        }

        LazyColumn(
            Modifier.padding(horizontal = 20.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp)
        ) {
            if (isPrimary) {
                item { PrimaryModeContent(state, viewModel) }
            } else {
                item { SecondaryModeContent(state, viewModel) }
            }

            // Status messages
            if (state is EnrollState.Error) {
                item {
                    Card(
                        Modifier.fillMaxWidth(),
                        shape = RoundedCornerShape(12.dp),
                        colors = CardDefaults.cardColors(containerColor = DangerDim)
                    ) {
                        Text(
                            (state as EnrollState.Error).message,
                            modifier = Modifier.padding(14.dp),
                            fontSize = 13.sp,
                            color = Danger
                        )
                    }
                }
            }

            if (state is EnrollState.Approved || state is EnrollState.JoinSuccess) {
                item {
                    Card(
                        Modifier.fillMaxWidth(),
                        shape = RoundedCornerShape(12.dp),
                        colors = CardDefaults.cardColors(containerColor = SuccessDim)
                    ) {
                        Text(
                            if (state is EnrollState.Approved) "Device approved and enrolled successfully"
                            else "Successfully joined pairing. Waiting for approval from primary device.",
                            modifier = Modifier.padding(14.dp),
                            fontSize = 13.sp,
                            color = Success
                        )
                    }
                }
            }

            item { Spacer(Modifier.height(20.dp)) }
        }
    }
}

@Composable
private fun PrimaryModeContent(state: EnrollState, viewModel: DeviceEnrollViewModel) {
    Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
        // Instructions
        Card(
            Modifier.fillMaxWidth(),
            shape = RoundedCornerShape(12.dp),
            colors = CardDefaults.cardColors(containerColor = AccentDim)
        ) {
            Text(
                "Start pairing to generate a code for the secondary device. The other device will need the pairing ID and challenge to join.",
                modifier = Modifier.padding(14.dp),
                fontSize = 12.sp,
                color = Accent,
                lineHeight = 18.sp
            )
        }

        when (state) {
            is EnrollState.Idle -> {
                Button(
                    onClick = { viewModel.initiatePairing() },
                    modifier = Modifier.fillMaxWidth(),
                    shape = RoundedCornerShape(12.dp),
                    colors = ButtonDefaults.buttonColors(containerColor = Accent)
                ) {
                    Text("Start Pairing", fontSize = 15.sp, fontWeight = FontWeight.SemiBold)
                }
            }
            is EnrollState.WaitingForSecondary -> {
                Card(
                    Modifier.fillMaxWidth(),
                    shape = RoundedCornerShape(12.dp),
                    colors = CardDefaults.cardColors(containerColor = BgCard)
                ) {
                    Column(Modifier.padding(16.dp)) {
                        Text("PAIRING ID", fontSize = 11.sp, color = TextTertiary)
                        Text(
                            state.pairingData.pairingId,
                            fontSize = 12.sp,
                            color = TextPrimary,
                            fontFamily = FontFamily.Monospace
                        )
                        Spacer(Modifier.height(8.dp))
                        Text("CHALLENGE", fontSize = 11.sp, color = TextTertiary)
                        Text(
                            state.pairingData.challenge,
                            fontSize = 12.sp,
                            color = TextPrimary,
                            fontFamily = FontFamily.Monospace
                        )
                        Spacer(Modifier.height(12.dp))
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            CircularProgressIndicator(Modifier.size(16.dp), color = Accent, strokeWidth = 2.dp)
                            Spacer(Modifier.width(8.dp))
                            Text("Waiting for secondary device...", fontSize = 13.sp, color = TextSecondary)
                        }
                    }
                }
            }
            is EnrollState.PairingJoined -> {
                Card(
                    Modifier.fillMaxWidth(),
                    shape = RoundedCornerShape(12.dp),
                    colors = CardDefaults.cardColors(containerColor = BgCard)
                ) {
                    Column(Modifier.padding(16.dp)) {
                        Text("Device wants to join:", fontSize = 13.sp, color = TextSecondary)
                        Text(
                            state.deviceName ?: "Unknown Device",
                            fontSize = 16.sp,
                            color = TextPrimary,
                            fontWeight = FontWeight.SemiBold
                        )
                    }
                }
                Spacer(Modifier.height(8.dp))
                Button(
                    onClick = { viewModel.approvePairing() },
                    modifier = Modifier.fillMaxWidth(),
                    shape = RoundedCornerShape(12.dp),
                    colors = ButtonDefaults.buttonColors(containerColor = Success)
                ) {
                    Text("Approve Device", fontSize = 15.sp, fontWeight = FontWeight.SemiBold)
                }
            }
            else -> {}
        }
    }
}

@Composable
private fun SecondaryModeContent(state: EnrollState, viewModel: DeviceEnrollViewModel) {
    var pairingId by remember { mutableStateOf("") }
    var challenge by remember { mutableStateOf("") }
    var deviceName by remember { mutableStateOf(android.os.Build.MODEL) }

    Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
        Card(
            Modifier.fillMaxWidth(),
            shape = RoundedCornerShape(12.dp),
            colors = CardDefaults.cardColors(containerColor = AccentDim)
        ) {
            Text(
                "Enter the pairing ID and challenge from the primary device to join.",
                modifier = Modifier.padding(14.dp),
                fontSize = 12.sp,
                color = Accent,
                lineHeight = 18.sp
            )
        }

        if (state is EnrollState.Idle || state is EnrollState.Error) {
            OutlinedTextField(
                value = pairingId,
                onValueChange = { pairingId = it },
                label = { Text("Pairing ID") },
                modifier = Modifier.fillMaxWidth(),
                singleLine = true,
                colors = OutlinedTextFieldDefaults.colors(
                    focusedBorderColor = Accent,
                    unfocusedBorderColor = TextTertiary,
                    focusedLabelColor = Accent,
                    cursorColor = Accent,
                    focusedTextColor = TextPrimary,
                    unfocusedTextColor = TextPrimary
                )
            )
            OutlinedTextField(
                value = challenge,
                onValueChange = { challenge = it },
                label = { Text("Challenge") },
                modifier = Modifier.fillMaxWidth(),
                singleLine = true,
                colors = OutlinedTextFieldDefaults.colors(
                    focusedBorderColor = Accent,
                    unfocusedBorderColor = TextTertiary,
                    focusedLabelColor = Accent,
                    cursorColor = Accent,
                    focusedTextColor = TextPrimary,
                    unfocusedTextColor = TextPrimary
                )
            )
            OutlinedTextField(
                value = deviceName,
                onValueChange = { deviceName = it },
                label = { Text("Device Name") },
                modifier = Modifier.fillMaxWidth(),
                singleLine = true,
                colors = OutlinedTextFieldDefaults.colors(
                    focusedBorderColor = Accent,
                    unfocusedBorderColor = TextTertiary,
                    focusedLabelColor = Accent,
                    cursorColor = Accent,
                    focusedTextColor = TextPrimary,
                    unfocusedTextColor = TextPrimary
                )
            )
            Button(
                onClick = { viewModel.joinPairing(pairingId, challenge, deviceName) },
                modifier = Modifier.fillMaxWidth(),
                enabled = pairingId.isNotBlank() && challenge.isNotBlank(),
                shape = RoundedCornerShape(12.dp),
                colors = ButtonDefaults.buttonColors(containerColor = Accent)
            ) {
                Text("Join Pairing", fontSize = 15.sp, fontWeight = FontWeight.SemiBold)
            }
        }
    }
}

File: ios/SsdidWallet/Feature/Device/DeviceEnrollScreen.swift
import SwiftUI
import SsdidCore

struct DeviceEnrollScreen: View {
    @Environment(AppRouter.self) private var router
    @EnvironmentObject private var services: ServiceContainer

    let keyId: String
    let mode: String // "primary" or "secondary"

    enum EnrollState {
        case idle
        case loading
        case waitingForSecondary(pairingId: String, challenge: String)
        case pairingJoined(pairingId: String, deviceName: String)
        case approved
        case joinSuccess
        case error(String)
    }

    @State private var state: EnrollState = .idle
    @State private var pairingIdInput = ""
    @State private var challengeInput = ""
    @State private var deviceNameInput = UIDevice.current.model
    @State private var pollingTask: Task<Void, Never>?

    private var isPrimary: Bool { mode == "primary" }

    private func makeDeviceManager() -> DeviceManager {
        DeviceManager(
            vault: services.vault,
            registryClient: HttpDeviceManagerRegistryClient(registryApi: services.httpClient.registry),
            ssdidClientProvider: SsdidClientDeviceProvider(ssdidClient: services.ssdidClient),
            deviceName: UIDevice.current.name,
            platform: "iOS"
        )
    }

    var body: some View {
        VStack(spacing: 0) {
            // Header
            HStack(spacing: 4) {
                Button {
                    pollingTask?.cancel()
                    router.pop()
                } label: {
                    Image(systemName: "chevron.left")
                        .foregroundStyle(Color.textPrimary)
                        .font(.system(size: 20))
                }
                .padding(.leading, 8)

                Text(isPrimary ? "Enroll New Device" : "Join as Device")
                    .font(.ssdidHeadline)
                    .foregroundStyle(Color.textPrimary)

                Spacer()
            }
            .padding(.vertical, 12)
            .padding(.trailing, 20)

            ScrollView {
                LazyVStack(spacing: 12) {
                    if isPrimary {
                        primaryModeContent
                    } else {
                        secondaryModeContent
                    }

                    // Error
                    if case .error(let message) = state {
                        Text(message)
                            .font(.system(size: 13))
                            .foregroundStyle(Color.danger)
                            .padding(14)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .background(Color.dangerDim)
                            .cornerRadius(12)
                    }

                    // Success states
                    if case .approved = state {
                        Text("Device approved and enrolled successfully")
                            .font(.system(size: 13))
                            .foregroundStyle(Color.success)
                            .padding(14)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .background(Color.successDim)
                            .cornerRadius(12)
                    }

                    if case .joinSuccess = state {
                        Text("Successfully joined pairing. Waiting for approval from primary device.")
                            .font(.system(size: 13))
                            .foregroundStyle(Color.success)
                            .padding(14)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .background(Color.successDim)
                            .cornerRadius(12)
                    }

                    Spacer().frame(height: 20)
                }
                .padding(.horizontal, 20)
            }
        }
        .background(Color.bgPrimary)
        .onDisappear {
            pollingTask?.cancel()
        }
    }

    @ViewBuilder
    private var primaryModeContent: some View {
        // Instructions
        Text("Start pairing to generate a code for the secondary device. The other device will need the pairing ID and challenge to join.")
            .font(.system(size: 12))
            .foregroundStyle(Color.ssdidAccent)
            .lineSpacing(4)
            .padding(14)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(Color.accentDim)
            .cornerRadius(12)

        switch state {
        case .idle:
            Button {
                Task { await initiatePairing() }
            } label: {
                Text("Start Pairing")
            }
            .buttonStyle(.ssdidPrimary)

        case .loading:
            ProgressView()
                .progressViewStyle(CircularProgressViewStyle(tint: Color.ssdidAccent))
                .frame(maxWidth: .infinity)
                .padding(.vertical, 20)

        case .waitingForSecondary(let pairingId, let challenge):
            VStack(alignment: .leading, spacing: 0) {
                Text("PAIRING ID")
                    .font(.system(size: 11))
                    .foregroundStyle(Color.textTertiary)
                Text(pairingId)
                    .font(.system(size: 12, design: .monospaced))
                    .foregroundStyle(Color.textPrimary)
                Spacer().frame(height: 8)
                Text("CHALLENGE")
                    .font(.system(size: 11))
                    .foregroundStyle(Color.textTertiary)
                Text(challenge)
                    .font(.system(size: 12, design: .monospaced))
                    .foregroundStyle(Color.textPrimary)
                Spacer().frame(height: 12)
                HStack(spacing: 8) {
                    ProgressView()
                        .progressViewStyle(CircularProgressViewStyle(tint: Color.ssdidAccent))
                        .scaleEffect(0.8)
                    Text("Waiting for secondary device...")
                        .font(.system(size: 13))
                        .foregroundStyle(Color.textSecondary)
                }
            }
            .ssdidCard()

        case .pairingJoined(_, let deviceName):
            VStack(alignment: .leading, spacing: 4) {
                Text("Device wants to join:")
                    .font(.system(size: 13))
                    .foregroundStyle(Color.textSecondary)
                Text(deviceName)
                    .font(.system(size: 16, weight: .semibold))
                    .foregroundStyle(Color.textPrimary)
            }
            .ssdidCard()

            Button {
                Task { await approvePairing() }
            } label: {
                Text("Approve Device")
                    .font(.ssdidBody.weight(.semibold))
                    .foregroundStyle(.white)
                    .frame(maxWidth: .infinity)
                    .padding(.vertical, 14)
                    .background(Color.success)
                    .cornerRadius(12)
            }

        default:
            EmptyView()
        }
    }

    @ViewBuilder
    private var secondaryModeContent: some View {
        // Instructions
        Text("Enter the pairing ID and challenge from the primary device to join.")
            .font(.system(size: 12))
            .foregroundStyle(Color.ssdidAccent)
            .lineSpacing(4)
            .padding(14)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(Color.accentDim)
            .cornerRadius(12)

        if case .idle = state {
            formFields
        } else if case .error = state {
            formFields
        }
    }

    @ViewBuilder
    private var formFields: some View {
        TextField("", text: $pairingIdInput,
                  prompt: Text("Pairing ID").foregroundStyle(Color.textTertiary))
            .textFieldStyle(.plain)
            .font(.ssdidMono)
            .foregroundStyle(Color.textPrimary)
            .padding(14)
            .background(Color.bgCard)
            .cornerRadius(12)
            .overlay(RoundedRectangle(cornerRadius: 12).stroke(Color.ssdidBorder, lineWidth: 1))

        TextField("", text: $challengeInput,
                  prompt: Text("Challenge").foregroundStyle(Color.textTertiary))
            .textFieldStyle(.plain)
            .font(.ssdidMono)
            .foregroundStyle(Color.textPrimary)
            .padding(14)
            .background(Color.bgCard)
            .cornerRadius(12)
            .overlay(RoundedRectangle(cornerRadius: 12).stroke(Color.ssdidBorder, lineWidth: 1))

        TextField("", text: $deviceNameInput,
                  prompt: Text("Device Name").foregroundStyle(Color.textTertiary))
            .textFieldStyle(.plain)
            .font(.ssdidBody)
            .foregroundStyle(Color.textPrimary)
            .padding(14)
            .background(Color.bgCard)
            .cornerRadius(12)
            .overlay(RoundedRectangle(cornerRadius: 12).stroke(Color.ssdidBorder, lineWidth: 1))

        Button {
            Task { await joinPairing() }
        } label: {
            Text("Join Pairing")
        }
        .buttonStyle(.ssdidPrimary(enabled: !pairingIdInput.isEmpty && !challengeInput.isEmpty))
        .disabled(pairingIdInput.isEmpty || challengeInput.isEmpty)
    }

    // MARK: - Actions

    private func initiatePairing() async {
        guard let identity = await services.vault.getIdentity(keyId: keyId) else {
            state = .error("Identity not found")
            return
        }

        state = .loading
        let deviceManager = makeDeviceManager()

        do {
            let pairingData = try await deviceManager.initiatePairing(identity: identity)
            state = .waitingForSecondary(pairingId: pairingData.pairingId, challenge: pairingData.challenge)

            // Start polling for secondary device joining
            pollingTask = Task {
                await pollPairingStatus(
                    deviceManager: deviceManager,
                    did: identity.did,
                    pairingId: pairingData.pairingId
                )
            }
        } catch {
            state = .error("Failed to initiate pairing: \(error.localizedDescription)")
        }
    }

    private func pollPairingStatus(deviceManager: DeviceManager, did: String, pairingId: String) async {
        // Poll every 3 seconds for up to 5 minutes
        let maxAttempts = 100
        for _ in 0..<maxAttempts {
            guard !Task.isCancelled else { return }

            do {
                try await Task.sleep(for: .seconds(3))
            } catch {
                return // Task cancelled
            }

            do {
                let status = try await deviceManager.checkPairingStatus(did: did, pairingId: pairingId)
                if status.status == "joined" || status.status == "pending_approval" {
                    await MainActor.run {
                        state = .pairingJoined(
                            pairingId: pairingId,
                            deviceName: status.deviceName ?? "Unknown Device"
                        )
                    }
                    return
                } else if status.status == "expired" || status.status == "rejected" {
                    await MainActor.run {
                        state = .error("Pairing session \(status.status)")
                    }
                    return
                }
            } catch {
                // Continue polling on transient errors
            }
        }

        await MainActor.run {
            state = .error("Pairing timed out. Please try again.")
        }
    }

    private func approvePairing() async {
        guard let identity = await services.vault.getIdentity(keyId: keyId) else {
            state = .error("Identity not found")
            return
        }

        guard case .pairingJoined(let pairingId, _) = state else { return }

        let deviceManager = makeDeviceManager()
        do {
            try await deviceManager.approvePairing(identity: identity, pairingId: pairingId)
            state = .approved
        } catch {
            state = .error("Failed to approve device: \(error.localizedDescription)")
        }
    }

    private func joinPairing() async {
        guard let identity = await services.vault.getIdentity(keyId: keyId) else {
            state = .error("Identity not found")
            return
        }

        state = .loading
        let deviceManager = makeDeviceManager()

        do {
            _ = try await deviceManager.joinPairing(
                did: identity.did,
                pairingId: pairingIdInput.trimmingCharacters(in: .whitespacesAndNewlines),
                challenge: challengeInput.trimmingCharacters(in: .whitespacesAndNewlines),
                identity: identity,
                deviceName: deviceNameInput.trimmingCharacters(in: .whitespacesAndNewlines)
            )
            state = .joinSuccess
        } catch {
            state = .error("Failed to join pairing: \(error.localizedDescription)")
        }
    }
}
